src/collector.rs

```diff
@@ -1,103 +1,175 @@
 // Copyright 2020 TiKV Project Authors. Licensed under Apache-2.0.
 
-#[derive(Debug, Clone)]
-pub(crate) struct SpanSet {
-    /// Span collection
-    pub spans: Vec<crate::Span>,
+use crossbeam::channel::Receiver;
 
-    /// Property collection
-    pub properties: crate::Properties,
-}
+use crate::trace::Span;
+use crate::utils::real_time_ns;
 
-pub(crate) struct CollectorInner {
-    start_time_ns: u64,
-    pub(crate) queue: crossbeam::queue::SegQueue<SpanSet>,
-    pub(crate) closed: std::sync::atomic::AtomicBool,
-}
+const INIT_LEN: usize = 1024;
+const INIT_BYTES_LEN: usize = 16384;
 
 pub struct Collector {
-    pub(crate) inner: std::sync::Arc<CollectorInner>,
+    rx: Receiver<SpanSet>,
+    trace_id: u64,
+    start_time_ns: u64,
+    start_time_cycles: u64,
 }
 
 impl Collector {
-    pub(crate) fn new(start_time_ns: u64) -> Self {
-        let collector = std::sync::Arc::new(crate::collector::CollectorInner {
-            start_time_ns,
-            queue: crossbeam::queue::SegQueue::new(),
-            closed: std::sync::atomic::AtomicBool::new(false),
-        });
-
-        crate::collector::Collector { inner: collector }
+    pub(crate) fn new(trace_id: u64, rx: Receiver<SpanSet>) -> Self {
+        Collector {
+            rx,
+            trace_id,
+            start_time_ns: real_time_ns(),
+            start_time_cycles: minstant::now(),
```

Comment:
The order is vital. `minstant::now()` should be called before `real_time_ns()`.

For initialization, `minstant::now()` may run 20ms for the first time. In that case, there will be a big gap between `start_time_ns` and `start_time_cycle`.

Reply:
All right. I see a `minstant::now()` in `start_trace()` which calls this fucntion.